archive/imo/imo1979_q1.lean

```diff
@@ -0,0 +1,302 @@
+`1+1/2+1/3+...+1/1319 - 2*(1/2+1/4+1/6+...+1/1318)=1/660+1/661+...+1/1319
+and now re-arranging as (1/660+1/1319)+(1/661+1/1318)+... we see that
+the numerator of each fraction is 1979 and the denominator is coprime to 1979
+(note that 1979 is prime). Hence the 1979-adic valuation of each fraction is positive,
+and thus the 1979-adic valuation of the sum is also positive.
+
+## Remarks on the formalisation
+
+The p-adic valuation function on ℚ is ℤ-valued and hence has v(0)=0 (rather than +∞);
+we thus had to occasionally deal with funny edge cases which aren't there mathematically.
+In retrospect it might have been easier to work with p-adic norms, which are ℚ-valued
+and don't have this problem.
+-/
+
+
+open_locale big_operators
+
+open finset
+
+open nat
+
+instance : fact (nat.prime 1979) := by norm_num
+
+open finset
+
+namespace imo1979q1
+
+-- some constants
+@[reducible] def a : ℚ := ∑ n in range 1320, (-1)^(n + 1) / n
+@[reducible] def b : ℚ := ∑ n in range 1320, 1 / n
+@[reducible] def c : ℚ := ∑ n in range 660, 1 / n
+@[reducible] def d : ℚ := ∑ n in Ico 660 1320, 1 / n
+@[reducible] def e : ℚ := ∑ n in range 330, 1 / (n + 660)
+@[reducible] def f : ℚ := ∑ n in range 330, 1 / (1319 - n)
+
+/-
+  The goal is equivalent to showing that the 1979-adic valuation of a is positive.
+
+  We start with some lemmas:
+
+  Lemma 1 : b - a = c
+  Lemma 2 : c + d = b
+  Corollary 3 : a = d
+  Lemma 4 : d = e + f
+
+  Then we prove the theorem, by showing the 1979-adic
+  valuation of e + f is positive.
+-/
+
+@[reducible] def double : ℕ ↪ ℕ := ⟨λ n, 2 * n, λ a b, mul_left_cancel' (by norm_num)⟩
+
+lemma lemma1 : b - a = c :=
+  calc b - a = ∑ n in range 1320, (1/n - (-1)^(n+1)/n) : by rw sum_sub_distrib
+    ... = ∑ n in range 1320, ite (even n) (2/n) 0 : by {
+      apply sum_congr rfl,
+      rintros x hx,
+      rw [pow_add, pow_one, mul_neg_one, neg_div, sub_neg_eq_add, ← _root_.add_div],
+      split_ifs,
+      { rcases h with ⟨n, rfl⟩,
+        rw [pow_mul, pow_two],
+        simp, norm_num },
+      { rw not_even_iff at h,
+        have h8 := nat.mod_add_div x 2,
+        rw h at h8,
+        suffices : (1 : ℚ) + (-1)^x = 0,
+          rw this, simp,
+        rw ← h8,
+        simp [pow_add, pow_mul],
+      }
+    }
+    ... = ∑ (x : ℕ) in filter even (range 1320), 2 / x : by rw sum_filter
+    ... = ∑ (x : ℕ) in map double (range 660), (2 : ℚ) / x : by {
+      apply sum_congr, swap, simp,
+      apply finset.subset.antisymm,
+      { intros x hx,
+        rw mem_filter at hx,
+        rcases hx with ⟨hx1, m, rfl⟩,
+        rw mem_map,
+        use m,
+        existsi _, refl,
+        rw mem_range at hx1 ⊢,
+        linarith },
+      { intros x hx,
+        rw mem_filter,
+        rw mem_map at hx,
+        rcases hx with ⟨a, ha, rfl⟩,
+        split,
+        { rw mem_range at ha ⊢,
+          change 2 * a < 1320,
+          linarith },
+        { existsi a,
+          refl
+        }
+      }
```

Comment:
Is it really this hard to prove that `filter even (range 1320) = map double (range 660)`? That's awful!

Reply:
I can do it a little faster with
```
example : filter even (range 1320) = map double (range 660) :=
begin
  ext a,
  simp only [mem_filter, mem_range, mem_map, double],
  split,
  { rintros ⟨x, m, rfl⟩,
    refine ⟨m, _, rfl⟩,
    rw mem_range,
    linarith },
  { rintros ⟨a, ha, rfl⟩,
    refine ⟨_, a, rfl⟩,
    rw mem_range at ha,
    change 2 * a < 1320,
    linarith, }
end
```

Reply:
In this case, the general form of the lemma seems to be easier to prove

```lean
example (n : ℕ) : filter even (range (2*n)) = map double (range n) :=
begin
  induction n with k IH,
  { refl },
  { rw [mul_succ, range_succ, range_succ, range_succ, map_insert, filter_insert, if_neg, 
        filter_insert, if_pos, IH],
    { refl },
    { simp with parity_simps },
    { simp with parity_simps } }
end
```

Reply:
I'd think a more general form takes a monotone function, filters an interval on membership of the image of that function and shows the result is the image of an interval given suitable conditions on the orders involved. With special cases for arithmetic progressions, i.e. images of intervals under functions `a * x + b` in `nat` or `int` being equivalent to filtering intervals on some modular arithmetic condition, and versions for `odd` and `even` being special cases of that. (And most of that making sense for both `set` and `finset`.) That may not make proofs significantly simpler, but it fits in with mathlib principles, and those more general lemmas about intervals, monotone functions and arithmetic progressions seem plausibly appropriate for mathlib proper.

Reply:
I've gone with @eric-wieser 's proof.

Reply:
Having said that, I now feel guilty because one of the main gains that we're getting from formalising IMO solutions is that it shows up gaps in the library...